src/promiseAllProperties.ts

```diff
@@ -1,12 +1,17 @@
-export type PromisesMap<T> = { [P in keyof T]: Promise<T[P]> | T[P] };
+type PlainObj = Record<string, unknown>;
+export type PromisesMap<T extends PlainObj> = {
+  [P in keyof T]: Promise<T[P]> | T[P];
+};
 
 /**
  * Receives an object with promise containing properties and returns a promise that resolves to an object
  * with the same properties containing the resolved values
  * @param  {PromisesMap<T>} promisesMap  the input object with a promise in each property
  * @return {Promise<T>}  a promise that resolved to an object with the same properties containing the resolved values
  */
-export default function promiseAllProperties<T>(promisesMap: PromisesMap<T>): Promise<T> {
+export default function promiseAllProperties<T extends PlainObj>(
+  promisesMap: PromisesMap<T>
+): Promise<T> {
   if (promisesMap === null || typeof promisesMap !== 'object') {
     return Promise.reject(new TypeError('The input argument must be of type Object'));
   }
```

Comment:
```suggestion
	if (
		!(typeof process !== undefined && process.env.NODE_ENV === 'production') &&
		(promisesMap === null ||
			typeof promisesMap !== 'object' ||
			Array.isArray(promisesMap))
	) {
		return Promise.reject(new TypeError('The input argument must be a plain object'));
	}
```

I'm tempted to also suggest something like this.

This adds a check to explicitly reject arrays but make the validation tree-shakeable in production.

Reply:
...this sort of change could be considered to be "breaking", but then again it could just as easily be argued that it is "not-really-breaking" since the correct behavior when the user passes an Array was never documented, and passing an array is hardly a reasonable usage of a function whose sole purpose for being is to allow people to not pass arrays in the first place.